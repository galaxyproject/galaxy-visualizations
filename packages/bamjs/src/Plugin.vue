<script setup>
import { BamFile } from "@gmod/bam";
import axios from "axios";
import { computed, onMounted, ref, watch } from "vue";

const props = defineProps({
    root: String,
    tracks: Array,
    datasetId: String,
    datasetUrl: String,
    settings: {
        type: Object,
        default: () => ({
            max_records: 100,
            region_start: 0,
            region_end: 10000,
        }),
    },
    specs: Object,
});

const loading = ref(false);
const error = ref(null);
const bamData = ref(null);
const headerInfo = ref(null);
const records = ref([]);

const maxRecords = computed(() => props.settings?.max_records || 100);
const regionStart = computed(() => props.settings?.region_start || 0);
const regionEnd = computed(() => props.settings?.region_end || 10000);

async function loadBamFile() {
    loading.value = true;
    error.value = null;

    try {
        if (props.datasetUrl.startsWith("mock://")) {
            headerInfo.value = {
                version: "1.6",
                sortOrder: "coordinate",
                references: [
                    { name: "chr1", length: 248956422 },
                    { name: "chr2", length: 242193529 },
                    { name: "chrX", length: 156040895 },
                ],
                readGroups: [{ id: "HG001", sample: "NA12878", library: "lib_HG001", platform: "ILLUMINA" }],
                programs: [
                    {
                        id: "bwa",
                        name: "bwa",
                        version: "0.7.17",
                        commandLine: "bwa mem -M -t 16 GRCh38.fa input.fastq",
                    },
                ],
            };

            records.value = [
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:1234:2000",
                    refName: "chr1",
                    start: 69091,
                    end: 69141,
                    cigar: "50M",
                    seq: "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATC",
                    qual: "JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ",
                    flags: 99,
                    mapq: 60,
                    tags: { AS: 50, XS: 10, NM: 0, MD: "50" },
                },
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:1234:2000",
                    refName: "chr1",
                    start: 69141,
                    end: 69191,
                    cigar: "50M",
                    seq: "CGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT",
                    qual: "IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIJ",
                    flags: 147,
                    mapq: 60,
                    tags: { AS: 50, XS: 8, NM: 0, MD: "50" },
                },
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:5678:2500",
                    refName: "chr2",
                    start: 158312,
                    end: 158362,
                    cigar: "25M1I24M",
                    seq: "GACTGCAGACTGCAGACTGCAGAACTGCAGACTGCAGACTGCAGACTGCAG",
                    qual: "HHHHHHHHHHHHHHHHHHHHHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGG",
                    flags: 0,
                    mapq: 42,
                    tags: { AS: 48, XS: 5, NM: 1, MD: "25^A24" },
                },
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:9012:3000",
                    refName: "chr1",
                    start: 52381,
                    end: 52426,
                    cigar: "45M5S",
                    seq: "TGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGC",
                    qual: "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF?????",
                    flags: 16,
                    mapq: 23,
                    tags: { AS: 40, XS: 15, NM: 2, MD: "20A10T14" },
                },
            ];

            return;
        }

        if (props.datasetUrl.startsWith("http")) {
            let bamFile;

            try {
                bamFile = new BamFile({
                    bamUrl: props.datasetUrl,
                    baiUrl: props.datasetUrl + ".bai",
                });

                headerInfo.value = await bamFile.getHeader();
            } catch (indexError) {
                console.warn("Index file not found, trying without index:", indexError.message);

                bamFile = new BamFile({
                    bamUrl: props.datasetUrl,
                });

                headerInfo.value = await bamFile.getHeader();
            }

            const refSeqs = headerInfo.value.references;
            if (refSeqs && refSeqs.length > 0) {
                const firstRef = refSeqs[0];
                const endPos = Math.min(regionEnd.value, firstRef.length);

                try {
                    const bamRecords = await bamFile.getRecordsForRange(firstRef.name, regionStart.value, endPos, {
                        maxRecords: maxRecords.value,
                    });

                    records.value = bamRecords;
                } catch (rangeError) {
                    console.warn("Could not get records for range:", rangeError.message);
                    records.value = [];
                }
            }
        } else {
            const response = await axios.get(props.datasetUrl);

            if (response.data.download_url) {
                const bamFile = new BamFile({
                    bamUrl: response.data.download_url,
                });

                headerInfo.value = await bamFile.getHeader();

                const refSeqs = headerInfo.value.references;
                if (refSeqs && refSeqs.length > 0) {
                    const firstRef = refSeqs[0];
                    const endPos = Math.min(regionEnd.value, firstRef.length);

                    const bamRecords = await bamFile.getRecordsForRange(firstRef.name, regionStart.value, endPos, {
                        maxRecords: maxRecords.value,
                    });

                    records.value = bamRecords;
                }
            } else {
                error.value = "Unable to access BAM file data";
            }
        }
    } catch (err) {
        error.value = `Error loading BAM file: ${err.message}`;
        console.error("BAM loading error:", err);
    } finally {
        loading.value = false;
    }
}

function formatRecord(record) {
    const flags = [];
    if (record.flags & 0x1) flags.push("paired");
    if (record.flags & 0x2) flags.push("proper_pair");
    if (record.flags & 0x4) flags.push("unmapped");
    if (record.flags & 0x8) flags.push("mate_unmapped");
    if (record.flags & 0x10) flags.push("reverse");
    if (record.flags & 0x20) flags.push("mate_reverse");
    if (record.flags & 0x40) flags.push("read1");
    if (record.flags & 0x80) flags.push("read2");
    if (record.flags & 0x100) flags.push("secondary");
    if (record.flags & 0x200) flags.push("qc_fail");
    if (record.flags & 0x400) flags.push("duplicate");
    if (record.flags & 0x800) flags.push("supplementary");

    return {
        name: record.name || "unknown",
        refName: record.refName || "unknown",
        start: record.start,
        end: record.end,
        cigar: record.cigar || "",
        seq: record.seq || "",
        qual: record.qual || "",
        flags: flags.join(","),
        mapq: record.mapq || 0,
        tags: record.tags
            ? Object.entries(record.tags)
                  .map(([k, v]) => `${k}:${v}`)
                  .join(";")
            : "",
    };
}

function formatHeader(header) {
    let output = "BAM Header Information:\n";
    output += "======================\n\n";

    if (header.version) {
        output += `Version: ${header.version}\n`;
    }

    if (header.sortOrder) {
        output += `Sort Order: ${header.sortOrder}\n`;
    }

    if (header.groupOrder) {
        output += `Group Order: ${header.groupOrder}\n`;
    }

    output += "\nReference Sequences:\n";
    output += "-------------------\n";
    if (header.references) {
        header.references.forEach((ref, index) => {
            output += `${index + 1}. ${ref.name} (length: ${ref.length})\n`;
        });
    }

    if (header.readGroups && header.readGroups.length > 0) {
        output += "\nRead Groups:\n";
        output += "------------\n";
        header.readGroups.forEach((rg, index) => {
            output += `${index + 1}. ID: ${rg.id}\n`;
            if (rg.sample) output += `   Sample: ${rg.sample}\n`;
            if (rg.library) output += `   Library: ${rg.library}\n`;
            if (rg.platform) output += `   Platform: ${rg.platform}\n`;
        });
    }

    if (header.programs && header.programs.length > 0) {
        output += "\nPrograms:\n";
        output += "---------\n";
        header.programs.forEach((prog, index) => {
            output += `${index + 1}. ID: ${prog.id}\n`;
            if (prog.name) output += `   Name: ${prog.name}\n`;
            if (prog.version) output += `   Version: ${prog.version}\n`;
            if (prog.commandLine) output += `   Command: ${prog.commandLine}\n`;
        });
    }

    return output;
}

function formatRecords(recordList) {
    if (!recordList || recordList.length === 0) {
        return "No records found in the specified region.";
    }

    let output = `\nBAM Records (showing ${recordList.length} records):\n`;
    output += "=".repeat(50) + "\n\n";

    recordList.forEach((record, index) => {
        const formatted = formatRecord(record);
        output += `Record ${index + 1}:\n`;
        output += `  Name: ${formatted.name}\n`;
        output += `  Reference: ${formatted.refName}\n`;
        output += `  Position: ${formatted.start}-${formatted.end}\n`;
        output += `  CIGAR: ${formatted.cigar}\n`;
        output += `  MAPQ: ${formatted.mapq}\n`;
        output += `  Flags: ${formatted.flags}\n`;
        if (formatted.seq) {
            output += `  Sequence: ${formatted.seq.substring(0, 100)}${formatted.seq.length > 100 ? "..." : ""}\n`;
        }
        if (formatted.qual) {
            output += `  Quality: ${formatted.qual.substring(0, 100)}${formatted.qual.length > 100 ? "..." : ""}\n`;
        }
        if (formatted.tags) {
            output += `  Tags: ${formatted.tags}\n`;
        }
        output += "\n";
    });

    return output;
}

watch(() => props.datasetUrl, loadBamFile);

onMounted(() => {
    if (props.datasetUrl) {
        loadBamFile();
    }
});
</script>

<template>
    <div class="bam-viewer">
        <div v-if="loading" class="loading">Loading BAM file...</div>

        <div v-else-if="error" class="error">
            {{ error }}
        </div>

        <div v-else-if="headerInfo || records.length > 0" class="bam-records">
            <div v-if="headerInfo">{{ formatHeader(headerInfo) }}</div>
            <div v-if="records.length > 0">{{ formatRecords(records) }}</div>
        </div>

        <div v-else class="loading">No BAM data available</div>
    </div>
</template>
