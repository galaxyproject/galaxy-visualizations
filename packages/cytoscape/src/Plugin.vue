<script setup lang="ts">
import Cytoscape, { type Core, type ElementDefinition, type ElementsDefinition } from "cytoscape";
import { onMounted, onUnmounted, ref, watch } from "vue";

import { parseSIF, runSearchAlgorithm, runTraversalType, styleGenerator } from "./utils";

const TEST_URL = "http://cdn.jsdelivr.net/gh/galaxyproject/galaxy-test-data/1.cytoscapejson";

type Settings = {
    directed: boolean;
    layout_name: "breadthfirst" | "circle" | "concentric" | "cose" | "grid" | "null" | "preset" | "random";
    curve_style: "haystack" | "straight" | "bezier" | "unbundled-bezier" | "segments" | "taxi";
    color_picker_nodes: string;
    color_picker_edges: string;
    color_picker_highlighted: string;
    graph_traversal?: "successors" | "predecessors" | "outgoers" | "incomers" | "roots" | "leaves";
    search_algorithm?: "bfs" | "dfs" | "kruskal" | "astar";
};

interface Props {
    root: string;
    datasetId: string;
    datasetUrl: string;
    settings: Settings;
}

const props = defineProps<Props>();

const viewport = ref(null);

const cytoscape = ref<Core | null>(null);
const dataset = ref<ElementsDefinition | ElementDefinition[]>();

async function render() {
    cytoscape.value = Cytoscape({
        container: viewport.value,
        elements: dataset.value,
        layout: {
            name: props.settings.layout_name,
            idealEdgeLength: () => 100,
            nodeOverlap: 20,
        },
        minZoom: 0.1,
        maxZoom: 50,
        style: styleGenerator(props.settings),
    });

    if (props.settings.search_algorithm === "kruskal") {
        //@ts-ignore
        cytoscape.value.elements().kruskal().edges().addClass("searchpath");
    }

    cytoscape.value.on("tap", "node", (e) => {
        const ele = e.target;

        const search_algorithm = props.settings.search_algorithm;
        const traversal_type = props.settings.graph_traversal;

        if (cytoscape.value && search_algorithm) {
            runSearchAlgorithm(cytoscape.value, ele.id(), search_algorithm);
        } else if (cytoscape.value && traversal_type) {
            runTraversalType(cytoscape.value, ele.id(), traversal_type);
        }
    });
}

async function getDataset() {
    try {
        const datasetUrl = props.datasetId === "__test__" ? TEST_URL : props.datasetUrl;
        const response = await fetch(datasetUrl);
        if (response.ok) {
            const content = await response.json();
            if (content.file_ext === "sif") {
                dataset.value = parseSIF(content).content;
            } else {
                dataset.value = content.elements ?? content;
            }
        } else {
            console.error("Fetch failed", response);
        }
    } catch (err) {
        console.error("Dataset fetch error", err);
    }
}

watch(
    () => props,
    () => render(),
    { deep: true },
);

onMounted(async () => {
    await getDataset();
    await render();
    window.addEventListener("resize", function () {
        cytoscape.value?.resize();
    });
});

onUnmounted(() => {
    window.removeEventListener("resize", function () {
        console.log("Removing resize event listener");
    });

    cytoscape.value?.destroy();
});
</script>

<template>
    <div class="h-screen overflow-hidden" ref="viewport"></div>
</template>
