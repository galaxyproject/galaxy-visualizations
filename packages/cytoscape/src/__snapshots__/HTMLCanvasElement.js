HTMLCanvasElement.prototype.toDataURL = vi.fn(() => "data:image/png;base64,MOCKDATA");

HTMLCanvasElement.prototype.getContext = vi.fn(function () {
    const canvas = this;
    if (!canvas._opsContainer) {
        const opsContainer = document.createElement("div");
        opsContainer.setAttribute("data-canvas-ops", "");
        canvas.insertAdjacentElement("afterend", opsContainer);
        canvas._opsContainer = opsContainer;
    }
    const opsContainer = canvas._opsContainer;
    function logOp(name, ...args) {
        const div = document.createElement("div");
        div.textContent = `${name}(${args.map((a) => JSON.stringify(a)).join(", ")})`;
        opsContainer.appendChild(div);
    }
    return {
        fillRect(x, y, w, h) {
            logOp("fillRect", x, y, w, h);
        },
        clearRect(x, y, w, h) {
            logOp("clearRect", x, y, w, h);
        },
        getImageData(x, y, w, h) {
            logOp("getImageData", x, y, w, h);
            return { data: new Uint8ClampedArray(w * h * 4) };
        },
        putImageData(imageData, x, y) {
            logOp("putImageData", x, y);
        },
        createImageData(w, h) {
            logOp("createImageData", w, h);
            return { data: new Uint8ClampedArray(w * h * 4) };
        },
        setTransform() {
            logOp("setTransform");
        },
        drawImage() {
            logOp("drawImage");
        },
        save() {
            logOp("save");
        },
        fillText(text, x, y) {
            logOp("fillText", text, x, y);
        },
        restore() {
            logOp("restore");
        },
        measureText() {
            logOp("measureText");
            return { width: 42 };
        },
        translate(x, y) {
            logOp("translate", x, y);
        },
        beginPath() {
            logOp("beginPath");
        },
        arc(x, y, r) {
            logOp("arc", x, y, r);
        },
        fill() {
            logOp("fill");
        },
        closePath() {
            logOp("closePath");
        },
        moveTo(x, y) {
            logOp("moveTo", x, y);
        },
        lineTo(x, y) {
            logOp("lineTo", x, y);
        },
        rotate(a) {
            logOp("rotate", a);
        },
        scale(x, y) {
            logOp("scale", x, y);
        },
        stroke() {
            logOp("stroke");
        },
        strokeRect(x, y, w, h) {
            logOp("strokeRect", x, y, w, h);
        },
        clip() {
            logOp("clip");
        },
    };
});
