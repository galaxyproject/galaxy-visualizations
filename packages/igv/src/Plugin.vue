<script setup lang="ts">
import { onMounted, onBeforeUnmount, ref, watch } from "vue";
import igv from "igv";
import { LastQueue } from "./lastQueue";
import { GalaxyApi, useDataTableStore, useDataJsonStore } from "galaxy-charts";
import CONFIG from "./config.yml";

const DEFAULT_TYPE = "annotation";
const DELAY = 500;
const URL = "https://s3.amazonaws.com/igv.org.genomes/genomes.json";

type Atomic = string | number | boolean | null | undefined;

interface Dataset {
    id: string;
    name: string;
}

interface Track {
    name: string;
    order?: number;
    [key: string]: TrackValue;
}

type TrackValue = Atomic | Dataset | Track;

const lastQueue = new LastQueue<any>(DELAY);

const props = defineProps<{
    datasetId: string;
    datasetUrl: string;
    root: string;
    settings: {
        locus: string;
        source: {
            from_igv: string;
            genome: any;
        };
    };
    specs: Record<string, unknown>;
    tracks: Track[];
}>();

// Emit events with TypeScript
const emit = defineEmits<{
    (event: "update", newSettings: any, newTracks: any): void;
}>();

const message = ref("");
const viewport = ref<HTMLDivElement | null>(null);

let igvBrowser: any = null;
let igvTracks: Array<Track> = [];
let parsedTracks: Array<Track> = [];

onMounted(() => {
    create();
});

onBeforeUnmount(() => {
    if (igvBrowser && typeof igvBrowser.dispose === "function") {
        igvBrowser.dispose();
    }
});

watch(
    () => props.tracks,
    () => lastQueue.enqueue(tracksParse, undefined, "tracksParse"),
    { deep: true, immediate: true },
);

watch(
    () => props.settings.source?.genome,
    () => lastQueue.enqueue(loadGenome, undefined, "loadGenome"),
);

watch(
    () => props.settings.locus,
    () => lastQueue.enqueue(locusSearch, undefined, "locusSearch"),
);

function getGenome() {
    const genome = props.settings.source.genome;
    if (genome) {
        if (props.settings.source.from_igv === "true") {
            return genome;
        } else {
            const genomeId = genome.id;
            const columns = genome.columns || [];
            const row = genome.row || [];
            const table = genome.table;
            const pathColumn = columns.indexOf("path");
            if (pathColumn >= 0 && row.length > pathColumn) {
                const fileName = row[pathColumn].split("/").pop();
                const apiPath = `${props.root}api/tool_data/${table}/fields/${genomeId}/files/`;
                if (table === "fasta_indexes") {
                    return {
                        id: genomeId,
                        fastaURL: `${apiPath}${fileName}`,
                        indexURL: `${apiPath}${fileName}.fai`,
                    };
                } else if (table === "twobit") {
                    return {
                        id: genomeId,
                        twoBitURL: `${apiPath}${fileName}`,
                    };
                }
            }
        }
    }
}

async function create() {
    if (props.datasetId && !props.settings.source.genome) {
        // access stores
        const dataTableStore = useDataTableStore();
        const dataJsonStore = useDataJsonStore();

        // collect dataset details
        const { data: dataset } = await GalaxyApi().GET(`/api/datasets/${props.datasetId}`);
        const dbkey = dataset.metadata_dbkey;

        // prep source
        let source = null;

        // attempt to match database key to fasta_indexes
        const dataTable = await dataTableStore.getDataTable("fasta_indexes");
        const dbkeyIndex = dataTable[0]?.value?.columns.indexOf("dbkey");
        const matchTable = dataTable.find(
            (item) => dbkeyIndex >= 0 && item.value?.row[dbkeyIndex] === dbkey
        );
        if (matchTable) {
            source = {
                from_igv: "false",
                genome: matchTable.value,
            };
        }

        // attempt to match database key to igv genomes
        if (!source) {
            const dataJson = await dataJsonStore.getDataJson(URL);
            const matchJson = dataJson.find((item) => item.value?.id === dbkey);
            if (matchJson) {
                source = {
                    from_igv: "true",
                    genome: matchJson.value,
                }
            }
        }

        // emit update
        const newSettings = source ? { source } : {};
        const newTracks = [{ urlDataset: { id: dataset.id } }];
        emit("update", newSettings, newTracks);
        console.error("[igv] Updating values.", newSettings, newTracks);
    } else {
        console.error("[igv] Skipping genome matching.");
    }
}

async function getDatasetType(datasetId: string): Promise<string | null> {
    try {
        const { data } = await GalaxyApi().GET(`/api/datasets/${datasetId}`);
        return data.extension;
    } catch (e) {
        message.value = `Failed to retrieve dataset type for: ${datasetId}`;
        console.error(message.value, e);
        return null;
    }
}

function getDatasetUrl(datasetId: string): string {
    return `${props.root}api/datasets/${datasetId}/display`;
}

function getIndexUrl(datasetId: string, format: string | null): string | null {
    const metadataKey = format ? CONFIG[format]?.index : null;
    return metadataKey ? `${props.root}api/datasets/${datasetId}/metadata_file?metadata_file=${metadataKey}` : null;
}

async function loadGenome() {
    const genomeConfig = getGenome();
    if (genomeConfig) {
        try {
            if (igvBrowser) {
                await igvBrowser.loadGenome(genomeConfig);
            } else if (viewport.value) {
                igvBrowser = await igv.createBrowser(viewport.value, { genome: genomeConfig });
            }

            // Refresh view
            await tracksLoad(true);
            await locusSearch();
            message.value = "";

            // Manually patch style, try to move this into css if possible
            const igvNavBar = igvBrowser.root.querySelector(".igv-navbar");
            if (igvNavBar) {
                Object.assign(igvNavBar.style, {
                    flexFlow: "column",
                    height: "unset",
                    paddingBottom: "3px",
                });
            }
        } catch (e) {
            message.value = "Failed to load genome.";
            console.error(message.value, e);
        }
    }
}

async function locusSearch() {
    const locus = props.settings.locus;
    if (igvBrowser) {
        if (!locus) {
            await igvBrowser.search("all");
        } else if (locusValid(locus)) {
            try {
                await igvBrowser.search(locus);
            } catch {
                console.warn("[igv] Invalid locus ignored:", locus);
            }
        }
    } else {
        message.value = "Failed to search.";
        console.error("[igv] Browser not available.");
    }
}

function locusValid(locus: string): boolean {
    const chrPattern = /^[\w.-]+$/;
    const rangePattern = /^[\w.-]+:\d{1,3}(,\d{3})*-\d{1,3}(,\d{3})*$/;
    return chrPattern.test(locus) || rangePattern.test(locus);
}

function trackHash(track: Track): string {
    const keys = Object.keys(track).sort();
    const str = JSON.stringify(track, keys);
    let hash = 2166136261;
    for (let i = 0; i < str.length; i++) {
        hash ^= str.charCodeAt(i);
        hash += (hash << 1) + (hash << 4) + (hash << 7) + (hash << 8) + (hash << 24);
    }
    return (hash >>> 0).toString(36);
}

function trackType(format: string | null): string {
    const normalized = format?.toLowerCase();
    if (normalized) {
        const config = CONFIG[normalized];
        if (config?.type) {
            return config.type;
        }
    }
    return DEFAULT_TYPE;
}

function tracksDiff(newTracks: Track[], existingTracks: Track[], force: boolean = false) {
    const existingMap = new Map(existingTracks.map((t) => [t.name, t]));
    const toRemove = existingTracks.filter((existing) => {
        const candidate = newTracks.find((t) => t.name === existing.name);
        return force || !candidate || trackHash(candidate) !== trackHash(existing);
    });
    const toAdd = newTracks.filter((t) => {
        const existing = existingMap.get(t.name);
        return force || !existing || trackHash(t) !== trackHash(existing);
    });
    return { toAdd, toRemove };
}

async function tracksLoad(force: boolean = false) {
    if (igvBrowser) {
        const newTracks = tracksResolve(parsedTracks);
        console.debug("[igv] Resolved Tracks", newTracks);
        const { toAdd, toRemove } = tracksDiff(newTracks, igvTracks, force);
        for (const old of toRemove) {
            const view = igvBrowser.trackViews.find((tv: any) => tv.track.name === old.name);
            if (view) {
                console.debug(`[igv] Removing existing track '${view.track.name}'`);
                igvBrowser.removeTrack(view.track);
            }
        }
        for (const t of toAdd) {
            try {
                console.debug(`[igv] Adding new track '${t.name}'`);
                await igvBrowser.loadTrack(t);
            } catch (e) {
                console.error(`[igv] Failed to load track '${t.name}'`, e);
            }
        }
        igvBrowser.reorderTracks();
        igvTracks = newTracks.map((t) => ({ ...t }));
    }
}

async function tracksParse() {
    parsedTracks = [];
    if (props.tracks) {
        for (const t of props.tracks) {
            // populate basic track parameters
            const dataset = t.urlDataset as Dataset;
            const displayMode = t.displayMode;
            const color = t.color;
            const format = dataset ? await getDatasetType(dataset.id) : null;
            const index = t.indexUrlDataset as Dataset;
            const name = t.name;
            const type = t.type && t.type !== "auto" ? t.type : trackType(format);
            const url = dataset ? getDatasetUrl(dataset.id) : null;
            // identify index source
            const indexURL = index ? getDatasetUrl(index.id) : getIndexUrl(dataset?.id, format);
            // create track configuration
            const trackConfig: Track = { color, displayMode, format, indexURL, name, type, url };
            console.debug("[igv] Track:", trackConfig);
            parsedTracks.push(trackConfig);
        }
    }
    tracksLoad();
}

function tracksResolve(rawTracks: Array<Track>) {
    const trackSeen = new Set<string>();
    const nameCounts = new Map<string, number>();
    const resolved: Array<Track> = [];
    rawTracks.forEach((track, index) => {
        const baseName = track.name || `track-${trackHash(track)}`;
        const trackKey = JSON.stringify({ ...track, name: baseName });
        const isValid = typeof track?.url === "string" && track.url.trim().length > 0;
        const isDuplicateConfig = trackSeen.has(trackKey);
        if (isValid && !isDuplicateConfig) {
            const count = nameCounts.get(baseName) ?? 0;
            let name = baseName;
            if (count > 0) {
                name = `${baseName} ~ ${count}`;
            }
            nameCounts.set(baseName, count + 1);
            trackSeen.add(trackKey);
            resolved.push({ ...track, name, order: index });
        } else if (!isValid) {
            console.warn(`[igv] Invalid track skipped: '${baseName}'`);
        } else {
            console.warn(`[igv] Duplicate track skipped: '${baseName}'`);
        }
    });
    return resolved;
}
</script>

<template>
    <div class="h-screen overflow-auto">
        <div v-if="message" class="bg-sky-100 border border-sky-200 p-1 rounded text-sky-800 text-sm">
            {{ message }}
        </div>
        <div ref="viewport"></div>
    </div>
</template>
