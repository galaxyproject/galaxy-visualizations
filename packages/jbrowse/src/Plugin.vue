<script setup>
import { onMounted, ref, watch } from "vue";
import { renderJBrowse } from "./render";

const props = defineProps({
    datasetId: String,
    datasetUrl: String,
    root: String,
    settings: Object,
    specs: Object,
    tracks: Array,
});

const viewport = ref(null);

function render() {
    const config = JSON.parse(JSON.stringify(props.settings.jbrowseConfig));
    if (viewport.value) {
        renderJBrowse(viewport.value, config);
    }
}

onMounted(() => {
    render();
});

watch(
    () => props,
    () => render(),
    { deep: true },
);
</script>

<template>
    <div ref="viewport"></div>
</template>
