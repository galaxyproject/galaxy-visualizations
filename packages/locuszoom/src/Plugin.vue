<script setup>
import { onMounted, watch, ref } from "vue";
import "locuszoom/dist/locuszoom.css";
import LocusZoom from "locuszoom/esm";
import LzTabixSource from "locuszoom/esm/ext/lz-tabix-source";
import { makeGWASParser } from "locuszoom/esm/ext/lz-parsers";
import { plotUpdatesUrl, plotWatchesUrl } from "locuszoom/esm/ext/lz-dynamic-urls";

LocusZoom.use(LzTabixSource);

window.nowrap = false;
window.mode = false;

const COLOR = "#E30A17";
const MAX_RANGE = 10000000;
const SIZE = 40;

const props = defineProps({
    datasetId: String,
    datasetUrl: String,
    root: String,
    settings: Object,
    specs: Object,
    tracks: Array,
});

const errorMessage = ref("");

function render() {
    const id = props.settings.tabix?.id;
    const chrIn = props.settings.chromosome;
    const startIn = props.settings.start;
    const endIn = props.settings.end;
    const chrom_col = props.settings.chrom_col;
    const pos_col = props.settings.pos_col;
    const ref_col = props.settings.ref_col;
    const alt_col = props.settings.alt_col;
    const pvalue_col = props.settings.pvalue_col;
    const is_neg_log_pvalue = props.settings.is_neg_log_pvalue;
    const beta_col = props.settings.beta_col;
    const stderr_beta_col = props.settings.stderr_beta_col;
    if (!id) {
        errorMessage.value = "Please select a Tabix file.";
        return;
    }
    if (endIn - startIn > MAX_RANGE) {
        errorMessage.value = "We cannot output more than 10Mb at a time!";
        return;
    }
    if (endIn <= startIn) {
        errorMessage.value = "Invalid input: End position must be bigger than start position!";
        return;
    }
    errorMessage.value = "";
    const gwasParser = makeGWASParser({
        chrom_col,
        pos_col,
        ref_col,
        alt_col,
        pvalue_col,
        is_neg_log_pvalue,
        beta_col,
        stderr_beta_col,
    });
    let data_sources = new LocusZoom.DataSources().add("assoc", [
        "TabixUrlSource",
        {
            url_data: `${props.root}api/datasets/${props.datasetId}/display`,
            url_tbi: `${props.root}api/datasets/${id}/display`,
            parser_func: gwasParser,
            overfetch: 0,
        },
    ]);
    let stateUrlMapping = { chr: "chrom", start: "start", end: "end", ldrefvar: "ld_variant" };
    let association_data_layer_mods = {
        id: "associationpvalues_",
        name: "association",
        point_shape: "circle",
        point_size: SIZE,
        color: COLOR,
        legend: [{ shape: "circle", color: COLOR, size: SIZE, label: "phen", class: "lz-data_layer-scatter" }],
        tooltip: {
            closable: true,
            show: { or: ["highlighted", "selected"] },
            hide: { and: ["unhighlighted", "unselected"] },
            html: `<strong>phen</strong><br><strong>{{assoc:variant|htmlescape}}</strong><br>
                P Value: <strong>{{assoc:log_pvalue|logtoscinotation|htmlescape}}</strong><br>
                Ref. Allele: <strong>{{assoc:ref_allele|htmlescape}}</strong><br>`,
        },
    };
    const data_layer = {
        ...LocusZoom.Layouts.get("data_layer", "association_pvalues", association_data_layer_mods),
        namespace: { assoc: "assoc" },
        data_operations: [],
    };
    const panel = LocusZoom.Layouts.get("panel", "association", {
        data_layers: [data_layer],
    });
    let plot = LocusZoom.populate("#lz-plot", data_sources, {
        state: { chr: chrIn, start: startIn, end: endIn },
        responsive_resize: true,
        panels: [panel],
        toolbar: LocusZoom.Layouts.get("toolbar", "standard_plot"),
    });
    plotUpdatesUrl(plot, stateUrlMapping);
    plotWatchesUrl(plot, stateUrlMapping);

    // manually set size of svg container
    document.querySelector("#lz-plot_svg").style.height = "calc(100vh - 50px)";
}

onMounted(() => {
    render();
});

watch(
    () => props,
    () => {
        render();
    },
    { deep: true },
);
</script>

<template>
    <div id="lz-plot" class="h-100"></div>
    <div v-if="errorMessage" class="mt-4 p-3 border bg-red-100 text-red-700 text-sm text-center rounded">
        {{ errorMessage }}
    </div>
</template>

<style>
.lz-panel-toolbar {
    display: none;
}
</style>
