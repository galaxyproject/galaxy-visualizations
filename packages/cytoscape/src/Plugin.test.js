import { mount, flushPromises } from "@vue/test-utils";
import { describe, expect, it, vi } from "vitest";
import Plugin from "./Plugin.vue";
import "./__snapshots__/HTMLCanvasElement.js";

const content = {
    data: {
        format_version: "1.0",
        generated_by: "cytoscape-3.1.1",
        elements: {
            nodes: [{ data: { id: "n0", name: "node0" } }, { data: { id: "n1", name: "node1" } }],
            edges: [{ data: { source: "n0", target: "n1" } }],
        },
        data: {
            name: "mock",
        },
    },
};

vi.spyOn(global, "fetch").mockResolvedValue({
    ok: true,
    json: async () => content,
});

describe("Plugin", () => {
    it("should render correctly", async () => {
        const mountPoint = document.createElement("div");
        document.body.appendChild(mountPoint);
        const wrapper = mount(Plugin, {
            attachTo: mountPoint,
            props: {
                root: "/",
                datasetId: "MY_DATASET_ID",
                datasetUrl: "MY_DATASET_URL",
                settings: {
                    curve_style: "haystack",
                    layout_name: "preset",
                    directed: false,
                    search_algorithm: undefined,
                    graph_traversal: undefined,
                    color_picker_nodes: "#548DB8",
                    color_picker_edges: "#A5A5A5",
                    color_picker_highlighted: "#C00000",
                },
            },
        });
        await flushPromises();
        await wrapper.vm.$nextTick();
        const html = document.body.innerHTML.replace(/__file=".*\/packages/g, '__file="<root>/packages');
        expect(html).toMatchSnapshot();
    });
});
