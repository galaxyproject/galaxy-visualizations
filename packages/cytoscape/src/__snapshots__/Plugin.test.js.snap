// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Plugin > should render correctly 1`] = `"<div><div data-v-app=""><div class="h-screen overflow-hidden __________cytoscape_container"><div style="position: relative; z-index: 0; overflow: hidden;"><canvas style="user-select: none; outline-style: none; position: absolute; z-index: 3;" data-id="layer0-selectbox" width="0" height="0"></canvas><canvas style="user-select: none; outline-style: none; position: absolute; z-index: 2;" data-id="layer1-drag" width="0" height="0"></canvas><canvas style="user-select: none; outline-style: none; position: absolute; z-index: 1;" data-id="layer2-node" width="0" height="0"></canvas></div></div></div></div>"`;
